vllm/model_executor/models/phi3v.py

```diff
@@ -330,27 +300,111 @@ def _calc_hd_transform_size(*, width: int, height: int, hd_num: int = 16):
-        # Temporary patch before dynamic number of image tokens is supported
-        _, _, h, w = ctx.get_multimodal_config().image_input_shape
-        if (w, h) != _calc_hd_transform_size(width=image.width,
-                                             height=image.height):
-            logger.warning(
-                "Dynamic image shape is currently not supported. "
-                "Resizing input image to (%d, %d).", w, h)
+def dummy_data_for_phi3v(ctx: InputContext, seq_len: int):
+    # Result in the max possible feature size (h:w = 16:1)
+    dummy_height, dummy_width = 8000, 50
+    image_feature_size = get_phi3v_image_feature_size(
+        ctx.get_hf_config(PretrainedConfig),
+        input_height=dummy_height,
+        input_width=dummy_width,
+    )
+
+    seq_data = dummy_seq_data_for_clip(
+        CLIP_VIT_LARGE_PATCH14_336_CONFIG,
+        seq_len,
+        image_token_id=32044,
+        image_feature_size_override=image_feature_size,
+    )
+    mm_data = dummy_pixel_data_for_clip(
+        CLIP_VIT_LARGE_PATCH14_336_CONFIG,
+        image_width_override=dummy_width,
+        image_height_override=dummy_height,
+    )
 
-            data.image = image.resize((w, h))
+    return seq_data, mm_data
 
-    return MULTIMODAL_REGISTRY._get_plugin_for_data_type(ImagePixelData) \
-            ._default_input_mapper(ctx, data)
 
+def input_processor_for_phi3v(ctx: InputContext, llm_inputs: LLMInputs):
+    multi_modal_data = llm_inputs.get("multi_modal_data")
+    if multi_modal_data is None or not isinstance(
+            multi_modal_data, (ImagePixelData, ImageFeatureData)):
+        return llm_inputs
 
-@MULTIMODAL_REGISTRY.register_image_pixel_input_mapper(_image_processor)
+    model_config = ctx.model_config
+    multimodal_config = ctx.get_multimodal_config()
+    hf_config = ctx.get_hf_config(PretrainedConfig)
+
+    if isinstance(multi_modal_data, ImagePixelData):
+        image = multi_modal_data.image
+        if isinstance(image, torch.Tensor):
+            _, _, _, h, w = image.shape
+        else:
+            w, h = image.size
+
+        w, h = _calc_hd_transform_size(width=w, height=h)
+
+        image_feature_size = get_phi3v_image_feature_size(hf_config,
+                                                          input_width=w,
+                                                          input_height=h)
+    else:
+        image_features = multi_modal_data.image_features
+        image_feature_size = image_features.shape[-2]
+
+    prompt_token_ids = llm_inputs["prompt_token_ids"]
+    tokenizer = cached_get_tokenizer(model_config.tokenizer)
+
+    # We need to get the token for "<", not "▁<"
+    # https://huggingface.co/microsoft/Phi-3-vision-128k-instruct/raw/main/tokenizer.json
+    a_token_id, = tokenizer.encode("a", add_special_tokens=False)
+    a_token_id_, *image_1_token_ids = tokenizer.encode(
+        "a<|image_1|>", add_special_tokens=False)
+    assert a_token_id == a_token_id_
```

Comment:
IMO we should just hardcode `image_1_token_ids` and leave a NOTE on it. The use of `<image_1>` is such a hack in the original model example.

Reply:
I left this code here so I can remember how to reconstruct the token representation for additional image tokens when we implement support for #5820 in the future.

Reply:
I can move this to a separate function to make it less confusing.